pkg/tls/certificatemanagement/ca.go

```diff
@@ -0,0 +1,596 @@
+	// GetOrCreateKeyPair returns a KeyPair. If one exists, some checks are performed. Otherwise, a new KeyPair is created.
+	GetOrCreateKeyPair(cli client.Client, secretName, secretNamespace string, dnsNames []string) (KeyPair, error)
+	// GetCertificate returns a Certificate. If the certificate is not found a k8s.io NotFound error is returned.
+	GetCertificate(cli client.Client, secretName, secretNamespace string) (Certificate, error)
+	// Issued returns true if the provided certificate was signed by CertificateManager and returns false for user provided KeyPairs or Certificate Management.
+	Issued(certificate Certificate) bool
+	AddToStatusManager(manager status.StatusManager, namespace string)
+	KeyPair
+}
+
+// KeyPair wraps a Secret object that contains a private key and a certificate. Whether CertificateManagement is
+// configured or not, KeyPair returns the right InitContainer, VolumeMount or Volume (when applicable).
+type KeyPair interface {
+	//UseCertificateManagement returns true if this key pair was not user provided and certificate management has been configured.
+	UseCertificateManagement() bool
+	// BYO returns true if this KeyPair was provided by the user. If BYO is true, UseCertificateManagement is false.
+	BYO() bool
+	InitContainer(namespace string) corev1.Container
+	VolumeMount() corev1.VolumeMount
+	VolumeMountKeyFilePath() string
+	VolumeMountCertificateFilePath() string
+	Volume() corev1.Volume
+	Certificate
+}
+
+// Certificate wraps the certificate. Combine this with a TrustedBundle, to mount a trusted certificate bundle to a pod.
+type Certificate interface {
+	X509Certificate() *x509.Certificate
+	Secret(namespace string) *corev1.Secret
+	HashAnnotationKey() string
+	HashAnnotationValue() string
+}
+
+// TrustedBundle is used to create a trusted certificate bundle of the CertificateManager CA and 0 or more Certificates.
+type TrustedBundle interface {
+	MountPath() string
+	ConfigMap(namespace string) *corev1.ConfigMap
+	HashAnnotations() map[string]string
+	VolumeMount() corev1.VolumeMount
+	Volume() corev1.Volume
+	//AddPEM adds a PEM block to the certificate bundle.
+	AddPEM(annotationName string, pem []byte)
+	AddCertificates(certificates ...Certificate)
+}
+
+var (
+	log                     = logf.Log.WithName("tls")
+	errInvalidCertDNSNames  = errors.New("cert has the wrong DNS names")
+	errInvalidCertNoPEMData = errors.New("cert has no PEM data")
+)
+
+type trustedBundle struct {
+	certificateManager CertificateManager
+	pem                []byte
+	annotations        map[string]string
+}
+
+// CreateTrustedBundle creates a TrustedBundle, which provides standardized methods for mounting a bundle of certificates to trust.
+func CreateTrustedBundle(ca CertificateManager, certificates ...Certificate) TrustedBundle {
+	secret := ca.Secret("")
+	bundle := &trustedBundle{certificateManager: ca,
+		pem: []byte(fmt.Sprintf("# certificate name: %s\n%s\n\n", secret.Name, string(secret.Data[corev1.TLSCertKey]))),
```

Comment:
So you have an interface name "Certificate", but you need to get the Secret from it to get a Key? I think there's a smell here either with the Certificate interface or the interface that's required for a trust bundle (as it's clearly not a certificate that's required here).